### DISCUSSION TITLE
Switch to Last / Previous Workspace - Solution

### DISCUSSION THREAD

**Discussion by tjex** [wezterm/wezterm#5833] [Show and tell]
Finally scratched that itch. And it feels so good.

Basically, when changing to a workspace, a `previous_workspace` field in `wezterm.GLOBAL` is updated with the current workspace. The function `switch_to_previous_workspace()` then accesses that field when being executed.

There are two functions needed for this to work. One for toggling between the last workspaces. And one for updating the previous workspace when we change workspaces via other means than toggling between them.

[My full wezterm conf](https://github.com/tjex/wezterm-conf)

## Solution

[see in full context](https://github.com/tjex/wezterm-conf/blob/08aa565f24171225ab8eda522a7c13be9663766d/functions/funcs.lua#L35-L59)

```lua
-- funcs.lua
M.switch_workspace = function(window, pane, workspace)
	local current_workspace = window:active_workspace()
	if current_workspace == workspace then
		return
	end


	window:perform_action(
		act.SwitchToWorkspace({
			name = workspace,
		}),
		pane
	)
	wezterm.GLOBAL.previous_workspace = current_workspace
end


M.switch_to_previous_workspace = function(window, pane)
	local current_workspace = window:active_workspace()
	local workspace = wezterm.GLOBAL.previous_workspace


	if current_workspace == workspace or wezterm.GLOBAL.previous_workspace == nil then
		return
	end


	M.switch_workspace(window, pane, workspace)
end

```

## Keybinds

A keybind would be set like this:

```lua
-- keybind.lua

local func = require("func")

-- ...
		{
			key = "l",
			mods = "ALT",
			action = wezterm.action_callback(function(window, pane)
				func.switch_to_last_workspace(window, pane)
			end),
		},
```

## Note

What's essential, is that any calls to `wezterm.action.SwitchToWorkspace` need to be replaced with the `switch_workspace()` function, which is effectively a wrapper around this call along with the `wezterm.GLOBAL.previous_workspace` logic. 

```lua
		{
			key = "2",
			mods = "ALT",
			action = wezterm.action_callback(function(window, pane)
				func.switch_workspace(window, pane, "admin") -- switch to 'admin' workspace
			end),
		},
```

Or if that's too clunky to bake into another functionality, you can also update the `previous_workspace` field before switching workspaces, like here in the sessionizer function ([discussion here](https://github.com/wez/wezterm/discussions/4796)).

[see in full context](https://github.com/tjex/dotfiles/blob/e38a3674bec9acb3bf678d038935add265e1fbbf/.config/wezterm/functions/sessioniser.lua)

```lua
-- ... more code
	-- update previous_workspace before changing to new workspace.
	wezterm.GLOBAL.previous_workspace = window:active_workspace()
	window:perform_action(
		act.InputSelector({
			action = wezterm.action_callback(function(win, _, id, label)
				if not id and not label then
					wezterm.log_info("Cancelled")
				else
					wezterm.log_info("Selected " .. label)
					win:perform_action(
						act.SwitchToWorkspace({
							name = id,
							spawn = { cwd = home .. label },
						}),
						pane
					)
				end
			end),
			fuzzy = true,
			title = "Select project",
			choices = projects,
		}),
		pane
	)

-- ... more code
```

**Comment by krispouille** [comment]
@tjex wow nice work ^^
I had a solution that I was not proud of; so I did not share. But I'll definitly rely on your solution.
Thanks a lot

**Comment by tjex** [comment]
@krispouille Found a few things in need of tweaking. Here is a non-permalink [link to main](https://github.com/tjex/wezterm-conf/blob/main/functions/funcs.lua) for the up to date state.
